### ISSUE TITLE
[linter] no_await_no_async

### ISSUE THREAD

**Issue by stephane-archer** [dart-lang/sdk#59814]
if the last await is deleted from a function, the function should not be async

``` Dart
void foo(String bar) async { // should ask for deleting the async keyword
// await Future.delay(Duration(second: 2));
print(bar);
}
```

**Comment by scheglov** [comment]
Interesting. I hoped to use such or similar lint to automatically find places in the analyzer itself where we have async methods. But it seems that this is not trivial.

If the return type is `void`, then there is no issues.

But if the return type is a `Future`, we cannot just change the return type, because the client might pass it where a `Future` is required. This requires global analysis, so is not a good fit for the current linter approach, which is local, at most at library level.

**Comment by scheglov** [comment]
https://dart-review.googlesource.com/c/sdk/+/403320

**Comment by scheglov** [comment]
My current approach is to report when a fix can be done without changing the type.

In a separate discussion @srawlins pointed out that we can also support `Future<T>` return types when all returns return `Future`, and never values. I will update my CL correspondingly.

**Comment by bwilkerson** [comment]
And my comments on the CL are to the effect that I think the lint should report in all cases, irrespective of whether we can write an automated fix for it. I can't think of any code that involves using an `async` modifier on a function / method whose body doesn't contain an `await` where removing the `async` (and possibly rewriting other portions of the function / method) isn't either possible or desirable. In other words, I can't think of a false positive that we need to avoid. Am I missing a use case?

**Comment by scheglov** [comment]
Rewriting
```dart
Future<int> f() async {
  return 0;
}
```
into
```dart
Future<int> f() {
  return Future.value(0);
}
```
is possible, but not desirable.

**Comment by srawlins** [comment]
Yeah I can see how replacing `return 0;` with `return Future.value(0);` is not a slam dunk. I can see how devs would prefer the former (with `async`), if there was no performance cost incurred by writing `async`. Personally, I can see myself preferring the latter, after some getting used to. But if you have a function that _must_ return `Future<int>`, and the body of the function has a half dozen int literals that it returns, it would be super tedious to write `Future.value(...)` over and over.

**Comment by FMorschel** [comment]
I'd see that happening (the preference for `0` over `Future.value(0)`) most of the time with overriden methods only, if not I'd actually prefer the lint to trigger, but I think it is another use-case.

If I could have a _nudge_ (discussed at https://github.com/dart-lang/sdk/issues/58862) telling me the return type could be changed I'd really like that. So that for functions that are used often that were once async for some depency that is not needed anymore can have a small performance gain (we could start with deprecation but that would be easy to do) or even for helper functions used sporadically that can easily be refactored.

This, I guess, would be another issue, I'll open it later for a future discussion if anyone likes the idea.

**Comment by scheglov** [comment]
<details>

<summary>Benchmark code</summary>

```dart
Future<int> getValueAsync() async {
  return 42;
}

Future<int> getValueFuture() {
  return Future.value(42);
}

int getValueSync() {
  return 42;
}

Future<void> runAsyncBenchmark() async {
  final timer = Stopwatch()..start();
  var acc = 0;
  for (var i = 0; i < 10000000; i++) {
    acc = (acc + await getValueAsync()) & 0xFFFFFF;
  }
  timer.stop();
  print(' Async Benchmark: ${timer.elapsedMicroseconds} us');
}

Future<void> runFutureBenchmark() async {
  final timer = Stopwatch()..start();
  var acc = 0;
  for (var i = 0; i < 10000000; i++) {
    acc = (acc + await getValueFuture()) & 0xFFFFFF;
  }
  timer.stop();
  print('Future Benchmark: ${timer.elapsedMicroseconds} us');
}

void main() async {
  // Warm-up (very basic in this example)
  print('--- warm-up');
  await runAsyncBenchmark();
  await runFutureBenchmark();

  // Run benchmarks for measurement
  print('--- benchmark');
  await runAsyncBenchmark();
  await runFutureBenchmark();
}
```

</details>


Results
```
--- warm-up
 Async Benchmark: 903326 us
Future Benchmark: 1491952 us
--- benchmark
 Async Benchmark: 881653 us
Future Benchmark: 1482963 us
```

It looks that `async` is faster than creating explicit `Future`.

**Comment by scheglov** [comment]
When we *can* change the return type, it would be really good to do so.

<details>
<summary>Benchmark code</summary>

```dart
Future<int> getValueAsync() async {
  return 42;
}

int getValueSync() {
  return 42;
}

Future<void> runAsyncBenchmark() async {
  final timer = Stopwatch()..start();
  var acc = 0;
  for (var i = 0; i < 10000000; i++) {
    acc = (acc + await getValueAsync()) & 0xFFFFFF;
  }
  timer.stop();
  print('Async Benchmark: ${timer.elapsedMicroseconds} us');
}

Future<void> runSyncBenchmark() async {
  final timer = Stopwatch()..start();
  var acc = 0;
  for (var i = 0; i < 10000000; i++) {
    acc = (acc + getValueSync()) & 0xFFFFFF;
  }
  timer.stop();
  print(' Sync Benchmark: ${timer.elapsedMicroseconds} us');
}

void main() async {
  // Warm-up (very basic in this example)
  print('--- warm-up');
  await runAsyncBenchmark();
  await runSyncBenchmark();

  // Run benchmarks for measurement
  print('--- benchmark');
  await runAsyncBenchmark();
  await runSyncBenchmark();
}
```

</details>

Results
```
--- warm-up
Async Benchmark: 895403 us
 Sync Benchmark: 3889 us
--- benchmark
Async Benchmark: 888209 us
 Sync Benchmark: 3973 us
```

It looks more than 100x faster for this trivial code.

**Comment by FMorschel** [comment]
I'd also like to mention https://github.com/dart-lang/language/issues/870 related to the return of `Future`s without `await`. JIC anyone finds this discussion in the future.

**Comment by incendial** [comment]
> https://dart-review.googlesource.com/c/sdk/+/403320

Hi @scheglov, if you don't mind, here are a few missing test cases that can help avoid early false positives:

```dart
// 1, same applies to IfElement
Future<void> function() async {
  final stream = Stream.fromIterable(['1', '2', '3', '4']);

  await for (var s in stream) {
    print(s);
  }
}

// 2, AFAIK async won't be highlighted right now, but should be
Future<String> anotherAsyncMethod() async => Future.value('value');

// 3, the body has implicit return (useFunction copied from your test cases)
useFunction(() async {
  if (condition) return Future.value('value');
});

// 4, removing async will make the return type Future<String?>?
Future<String?> get(String? x) async => x == null ? null : Future.value(x);
```

**Comment by scheglov** [comment]
Thank you, @incendial.

Yes, I found (3) while working https://dart-review.googlesource.com/c/sdk/+/403082

And I forgot (1) about `await for`, you probably mean `ForElement`, not `IfElement`?

And we need support for `ExpressionBody`, so (2).

And we need to check for nullable `Future` expression type vs. non-nullable `Future` return type (4).

**Comment by incendial** [comment]
> And I forgot (1) about await for, you probably mean ForElement, not IfElement?

Yep, that's correct

**Comment by FMorschel** [comment]
I think this may be the same as https://github.com/dart-lang/sdk/issues/58901?

**Comment by bwilkerson** [comment]
> Rewriting ... into ... is possible, but not desirable.

I think it's important to separate the question of which cases the lint should flag and the question of which fixes we should automate.

Are you suggesting that
```dart
Future<int> f() async {
  return 0;
}
```
shouldn't be flagged by the lint, or that the one possible fix that I mentioned shouldn't be implemented?

**Comment by scheglov** [comment]
I think for code
```dart
Future<int> f() async {
  return 0;
}
```
we don't want to report the lint because we don't know how `f` is used, maybe the result is passed to a place that wants `Future<int>`. So, reporting the lint for such cases will produce a lot of diagnostics that the user not necessary can do anything about.

But an Quick Assist to convert into
```dart
int f() {
  return 0;
}
```
would be fine, because we expect that the user knows what he is doing, and will update any follow-up errors, or revert the change.

**Comment by scheglov** [comment]
https://dart-review.googlesource.com/c/sdk/+/403320 should fix false positives and false negatives mentioned [above](https://github.com/dart-lang/sdk/issues/59814#issuecomment-2575821131).

**Comment by scheglov** [comment]
The CL landed.
Now we need a quick fix in DAS.

**Comment by lrhn** [comment]
(About the benchmark code above: One reason `async` is faster is probably that it knows the static type of the expression, so it can call the internal `_Future._completeWithValue` directly, where the `Future.value` constructor starts with an expensive `if (value is Future<T>)` test. It may also complete it synchronously when it knows it's a value, where `Future.value` always completes asynchronously, which means scheduling a microtask. If nobody listens to a synchronously completed future, it doesn't have to schedule a microtask at all. If you change the `Future.value` call to `(Completer<int>.sync()..complete(42)).future;` then it becomes faster than `async` in many cases, and only a little slower when AoT native compiled.)

Edit: The fastest way (so far) to create a future with a value is:
```dart
Future<int> getValueFuture(int v) {
  return Future<int>.sync(() => v);
}
```
even if the value isn't fixed and it has to create a closure.
It's faster than an async function.

**Comment by srawlins** [comment]
> The CL landed.
> Now we need a quick fix in DAS.

Are you keeping this issue open to track the quick fix?

**Comment by FMorschel** [comment]
I volunteer to work on this fix. If you prefer to create a new issue for it, just CC me so I can correctly reference it on the CL, thanks!

**Comment by FMorschel** [comment]
Here is the CL https://dart-review.googlesource.com/c/sdk/+/425602
